jina/optimizers/parameters.py

```diff
@@ -0,0 +1,181 @@
+from typing import Optional, Sequence, Union
+
+import inspect
+import ruamel.yaml
+
+from jina.helper import yaml
+
+
+class OptimizationParameter:
+    def __init__(
+        self,
+        parameter_name: str,
+        executor_name: str = None,
+        prefix: str = "JINA",
+        env_var: str = None,
+        method:str = None
+    ):
+        if env_var is None:
+            self.env_var = f"{prefix}_{executor_name}_{parameter_name}".upper()
+        else:
+            self.env_var = env_var
+        self.parameter_name = parameter_name
+
+    @classmethod
+    def to_yaml(cls, representer, data):
+        """Required by :mod:`ruamel.yaml.constructor` """
+        tmp = data._dump_instance_to_yaml(data)
+        representer.sort_base_mapping_type_on_output = False
+        return representer.represent_mapping("!" + cls.__name__, tmp)
+
+    @staticmethod
+    def _dump_instance_to_yaml(instance):
+
+        attributes = inspect.getmembers(instance, lambda a: not (inspect.isroutine(a)))
+        return {
+            a[0]: a[1]
+            for a in attributes
+            if not (a[0].startswith("__") and a[0].endswith("__"))
+        }
+
+    @classmethod
+    def from_yaml(cls, constructor, node):
+        """Required by :mod:`ruamel.yaml.constructor` """
+        return cls._get_instance_from_yaml(constructor, node)
+
+    @classmethod
+    def _get_instance_from_yaml(cls, constructor, node):
+        data = ruamel.yaml.constructor.SafeConstructor.construct_mapping(
+            constructor, node, deep=True
+        )
+        return cls(**data)
+
+
+class IntegerParameter(OptimizationParameter):
+    def __init__(
+        self,
+        low: int,
+        high: int,
+        step_size: int = 1,
+        log: bool = False,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.low = low
+        self.high = high
+        self.step_size = step_size
+        self.log = log
+        self.method = "suggest_int"
+
+    def to_optuna_args(self):
+        return {
+            "name": self.env_var,
+            "low": self.low,
+            "high": self.high,
+            "step": self.step_size,
+            "log": self.log,
+        }
+
+yaml.register_class(IntegerParameter)
+
+class FloatParameter(OptimizationParameter):
+    def __init__(
+        self,
+        low: float,
+        high: float,
+        step_size: Optional[float] = None,
```

Comment:
why is step_size for Float Optional and not for Integer?

I think that for these classes we will soon have an explosion. 

I would separate the concerns of defining the `HyperParamSpace (low, high...)` and the `Sampling` which defines `step_size, log ...` (like this you could have just one FloatParameter work with `UniformSampler` or with `StepSampler`)

Reply:
This mimics exactly the optuna behavior. When we have another sampler, which we can supply to optuna, we can think about abstracting this. I believe you think way to complicated here. This is a optimization, where you won't have thousands of thousands of executions but rather in the hundreds with a huge parameter set. Fine-grained control over the sampling will not make your life much easier.

Reply:
If the use is not ready okey, but I think when it may come to the point of trying different strategies with the same ParameterSpace, this abstraction can make a significant difference.